themes/default/content/blog/repairing-state-with-pulumi-refresh/index.md

```diff
@@ -0,0 +1,98 @@
+
+### Out of bound changes
+
+Pulumi expects to be the sole controller for the resources it provisions. An out of bound change happens when this principle is violated. This is best illustrated with an example. Imagine creating an S3 bucket with Pulumi:
+
+```typescript
+const bucket = new aws_native.s3.Bucket("my-bucket", {
+    tags: [{
+            key: "pulumi-bucket",
+            value: "not important",
+        }],
+});
+```
+
+Then someone goes into the aws console and deletes the bucket. What happens if you then go change a tag on the bucket:
+
+```typescript
+const bucket = new aws_native.s3.Bucket("my-bucket", {
+    tags: [{
+            key: "pulumi-bucket",
+            value: "don’t delete me",
+        }],
+});
+```
+
+Pulumi has recorded creating a bucket, but it does not know that the bucket was deleted. It will try to update the bucket and fail, since the bucket does not exist anymore.
+
+### Terminating Pulumi Mid-Operation
+
+Pulumi also gets confused if the CLI or Pulumi Provider process is terminated while an update is taking place. The Pulumi CLI checks in its state before and after each operation it performs. Creating the bucket from above goes something like:
+
+a. Started “pulumi up”.
+b. About to create `aws_native:s3:Bucket::my-bucket`.
+c. Created `aws_native:s3:Bucket::my-bucket` with ID `my-bucket-6e3d099`.
+d. Finished “pulumi up”.
+
+If the Pulumi CLI process is terminated between steps (b) and (c), then Pulumi does not know if
+“my-bucket” exists or not. Unlike (1), Pulumi is aware that it does not know if the resource was created. This will be displayed as a pending CREATE operation. Depending on what type of operation was interrupted, you can see pending CREATE, UPDATE, DELETE, READ and IMPORT operations.
+
+![Pending Operations Warning](pending-ops-warning.png)
+
+## The Solution
+
+As of [#10394](https://github.com/pulumi/pulumi/pull/10394), the solution to both problems is [pulumi refresh](https://www.pulumi.com/docs/reference/cli/pulumi_refresh/). The `pulumi refresh` command modifies Pulumi&rsquo;s state so it matches the state of the underlying providers. Most discrepancies can be resolved automatically by checking Pulumi&rsquo;s state against the underlying resource provider. For our aws bucket example, Pulumi would ask the aws-native provider if bucket `my-bucket-6e3d099` and what tags it has.
+
+This works for all resources that Pulumi knows about. This set includes Pulumi managed resources that were changed manually, or for pending UPDATE, DELETE, READ and IMPORT operations. Pulumi does not know the resource ID for pending CREATE operations (it might not exist), so it can&rsquo;t resolve everything for you. It needs a little bit of help. For each pending CREATE, Pulumi will ask you what to do:
```

Comment:
```suggestion
This works for all resources that Pulumi knows about. This set includes Pulumi managed resources that were changed manually, or for pending UPDATE, DELETE, READ and IMPORT operations. Pulumi does not know the resource ID for pending CREATE operations (it might not exist), so it can't resolve everything for you. It needs a little bit of help. For each pending CREATE, Pulumi will ask you what to do:
```